### PR TITLE
ACRO parameter and EXPO update

### PR DESCRIPTION
**Pull request by lthall**
Use maximum rate parameters for ACRO in degrees per second
Update the expo function to provide greater range

### CODE REVIEW

File: ArduCopter/Parameters.cpp [hunk 1]
```diff
@@ -1069,6 +1067,36 @@ const AP_Param::GroupInfo ParametersG2::var_info[] = {
     AP_GROUPINFO("GUID_TIMEOUT", 46, ParametersG2, guided_timeout, 3.0),
 #endif
 
+#if MODE_ACRO_ENABLED == ENABLED
+    // @Param: ACRO_RP_RATE
+    // @DisplayName: Acro Roll and Pitch Rate
+    // @Description: Acro mode maximum roll and pitch rate.  Higher values mean faster rate of rotation
+    // @Range: 1 360
+    // @User: Standard
```
Comment (amilcarlucas) [inline ArduCopter/Parameters.cpp:1075]: Please add units

File: ArduCopter/Parameters.cpp [hunk 2]
```diff
@@ -1069,6 +1067,36 @@ const AP_Param::GroupInfo ParametersG2::var_info[] = {
...
+    AP_GROUPINFO("ACRO_Y_RATE", 48, ParametersG2, acro_y_rate, ACRO_Y_RATE_DEFAULT),
+#endif
+
+    // @Param: PILOT_Y_RATE
+    // @DisplayName: Pilot controlled yaw rate
+    // @Description: Pilot controlled yaw rate max.  Used in all pilot controlled modes except Acro
+    // @Range: 1 360
+    // @User: Standard
+    AP_GROUPINFO("PILOT_Y_RATE", 49, ParametersG2, pilot_y_rate, PILOT_Y_RATE_DEFAULT),
```
Comment (andyp1per) [inline ArduCopter/Parameters.cpp:1091]: You need pilot RP rate/expo as well?

File: ArduCopter/Parameters.cpp [hunk 3]
```diff
@@ -1069,6 +1067,36 @@ const AP_Param::GroupInfo ParametersG2::var_info[] = {
...
+
+    // @Param: PILOT_Y_EXPO
+    // @DisplayName: Pilot controlled yaw expo
+    // @Description: Pilot controlled yaw expo.  Used in all pilot controlled modes except Acrxo
```
Comment (andyp1per) [inline ArduCopter/Parameters.cpp:1095]: Spelling Acrox

File: ArduCopter/Parameters.h [hunk 4]
```diff
@@ -650,6 +650,14 @@ class ParametersG2 {
 #if MODE_GUIDED_ENABLED == ENABLED
     AP_Float guided_timeout;
 #endif
+
+#if MODE_ACRO_ENABLED == ENABLED
+    // Acro parameters
+    AP_Float                acro_rp_rate;
```
Comment (andyp1per) [inline ArduCopter/Parameters.h:656]: What about pilot rp?

File: ArduCopter/mode_acro.cpp [hunk 5]
```diff
@@ -96,46 +96,34 @@ float ModeAcro::throttle_hover() const
 
 // get_pilot_desired_angle_rates - transform pilot's roll pitch and yaw input into a desired lean angle rates
 // returns desired angle rates in centi-degrees-per-second
-void ModeAcro::get_pilot_desired_angle_rates(int16_t roll_in, int16_t pitch_in, int16_t yaw_in, float &roll_out, float &pitch_out, float &yaw_out)
+void ModeAcro::get_pilot_desired_angle_rates(float roll_in, float pitch_in, float yaw_in, float &roll_out, float &pitch_out, float &yaw_out)
 {
+    // todo: provide normalised commands from RC library so this can be removed.
+    roll_in /= ROLL_PITCH_YAW_INPUT_MAX;
```
Comment (andyp1per) [inline ArduCopter/mode_acro.cpp:102]: Can these be inverted so that its a multiply?

File: ArduCopter/mode_acro.cpp [hunk 6]
```diff
@@ -96,46 +96,34 @@ float ModeAcro::throttle_hover() const
...
+    // roll expo
+    rate_bf_request_cd.x = g2.acro_rp_rate * 100.0 * input_expo(roll_in, g.acro_rp_expo);
+
+    // pitch expo
+    rate_bf_request_cd.y = g2.acro_rp_rate * 100.0 * input_expo(pitch_in, g.acro_rp_expo);
 
-    // calculate yaw rate request
-    rate_bf_request.z = get_pilot_desired_yaw_rate(yaw_in);
+    // yaw expo
+    rate_bf_request_cd.z = g2.acro_rp_rate * 100.0 * input_expo(yaw_in, g2.acro_y_expo);
```
Comment (andyp1per) [inline ArduCopter/mode_acro.cpp:126]: Should be acro_y_rate

**Review by andyp1per — CHANGES_REQUESTED** [review]

File: ArduCopter/Parameters.cpp (hunk 2 above)
Comment (lthall) [reply ArduCopter/Parameters.cpp:1091]: The Pilot parameters are not used for everything other than ACRO. So no roll and pitch rates.

File: ArduCopter/Parameters.h (hunk 4 above)
Comment (lthall) [reply ArduCopter/Parameters.h:656]: There isn't one.

File: ArduCopter/config.h [hunk 7]
```diff
@@ -450,18 +450,14 @@
...
-#ifndef ACRO_YAW_P
- # define ACRO_YAW_P                4.5f
-#endif
-
 #ifndef ACRO_LEVEL_MAX_ANGLE
  # define ACRO_LEVEL_MAX_ANGLE      3000
 #endif
 
+#ifndef ACRO_LEVEL_MAX_OVERSHOOT
+ # define ACRO_LEVEL_MAX_OVERSHOOT  1000
```
[the remark this answers is not in the document]
Comment (lthall) [reply ArduCopter/config.h:458]: I am going with clarity over speed here.

File: ArduCopter/mode_acro.cpp (hunk 5 above)
Comment (lthall) [reply ArduCopter/mode_acro.cpp:102]: Yes but it should also be done back in the RC library and removed from here completly. So again I will stick with clarity.

File: ArduCopter/mode_acro.cpp (hunk 6 above)
Comment (lthall) [reply ArduCopter/mode_acro.cpp:126]: Legend!

**Review by andyp1per — APPROVED** [review]

File: ArduCopter/mode.cpp [hunk 8]
```diff
@@ -918,30 +918,22 @@ Mode::AltHoldModeState Mode::get_alt_hold_state(float target_climb_rate_cms)
 
 // transform pilot's yaw input into a desired yaw rate
 // returns desired yaw rate in centi-degrees per second
-float Mode::get_pilot_desired_yaw_rate(int16_t stick_angle)
+float Mode::get_pilot_desired_yaw_rate(float yaw_in)
 {
+    // todo: provide normalised commands from RC library so this can be removed.
+    yaw_in /= ROLL_PITCH_YAW_INPUT_MAX;
```
Comment (IamPete1) [inline ArduCopter/mode.cpp:924]: should be able to use `channel_yaw->norm_input()`

File: ArduCopter/Parameters.h [hunk 9]
```diff
@@ -650,6 +650,14 @@ class ParametersG2 {
 #if MODE_GUIDED_ENABLED == ENABLED
     AP_Float guided_timeout;
 #endif
+
+#if MODE_ACRO_ENABLED == ENABLED
```
Comment (IamPete1) [inline ArduCopter/Parameters.h:654]: these are not only used in acro

File: ArduCopter/Parameters.cpp [hunk 10]
```diff
@@ -1069,6 +1067,40 @@ const AP_Param::GroupInfo ParametersG2::var_info[] = {
     AP_GROUPINFO("GUID_TIMEOUT", 46, ParametersG2, guided_timeout, 3.0),
 #endif
 
+#if MODE_ACRO_ENABLED == ENABLED
```
Comment (IamPete1) [inline ArduCopter/Parameters.cpp:1070]: these are not only used in acro

**Comment by Hwurzburg** [comment]
what happens to present users of the params that are removed? or I am confused about ACRO_YAW_P,etc. ?

**Comment by IamPete1** [comment]
Just going to re-state that I think we should move this stuff into the attitude control library so we can share input shaping between plane and copter. We have acro mode (QACRO) on plane. And the yaw stuff is/should be used in all manual modes in anycase. 

Happy to help out with the param conversion.

File: ArduCopter/Parameters.cpp (hunk 1 above)
Comment (lthall) [reply ArduCopter/Parameters.cpp:1076]: Done

File: ArduCopter/Parameters.cpp (hunk 10 above)
Comment (lthall) [reply ArduCopter/Parameters.cpp:1070]: Nice catch!

File: ArduCopter/mode.cpp (hunk 8 above)
Comment (lthall) [reply ArduCopter/mode.cpp:924]: NICE!! I didn't know that got put in there.

**Comment by lthall** [comment]
> what happens to present users of the params that are removed? or I am confused about ACRO_YAW_P,etc. ?

We will add parameter conversions. I generally get help from Randy to do that.

**Comment by lthall** [comment]
> Just going to re-state that I think we should move this stuff into the attitude control library so we can share input shaping between plane and copter.

I am not sure what you are suggesting to put into the attitude control library. The expo and rates are acro specific and change for other modes. That is why we have both an ACRO and PILOT versions for yaw. I moved the expo calculation to the math library to make it easy to use anywhere.

The Attitude Library parameters are there to define how attitude commands are smoothed but limits are only aircraft limits. They are not intended to be command limitations.

**Comment by IamPete1** [comment]
> The expo and rates are acro specific

The yaw params are used in all pilot input modes.

IMHO the more code and params shared between copter and quadplane the better. Admittedly attitude control is not a great place, but I cant think of a better without making a new lib. 

1. quadplane gets copters improvements for for free, we don't have to put in the duplicated effort to bring them across.
2. Users already understand the params and ranges when they move from copter to quadplane, (or the other way of course, but that is less common) 

Ideally I would want the whole normalized input to rate/angle output in the lib. 

Maybe a pilot input library. We could move all the `PILOT_` and `ACRO_` params allong with the stick input to rate/angle/climbrate functions.

Happy to do the legwork on this one.

**Comment by lthall** [comment]
> Maybe a pilot input library. We could move all the PILOT_ and ACRO_ params along with the stick input to rate/angle/climb rate functions.

This is a Randy question I think. It would be a good time to change over to degrees and m/s ect for any of those parameters.

File: libraries/AP_Math/control.cpp [hunk 11]
```diff
@@ -407,3 +407,15 @@ float kinematic_limit(Vector3f direction, float max_xy, float max_z_pos, float m
     }
     return fabsf(max_z_neg/direction.z);
 }
+
+// calculate the expo function on the normalised input
+// input must be in the range of -1 to 1
+// expo should be less than 1.0 but limited to be less than 0.95
+float input_expo(float input, float expo)
+{
+    if (expo < 0.95) {
```
Comment (Gone4Dirt) [inline libraries/AP_Math/control.cpp:416]: Whats the reason for 0.95 and not 0.99?

File: libraries/AP_Math/control.cpp [hunk 12]
```diff
@@ -407,3 +407,15 @@ float kinematic_limit(Vector3f direction, float max_xy, float max_z_pos, float m
...
+        return (1 - expo) * input / (1 - expo * fabsf(input));
```
Comment (Gone4Dirt) [inline libraries/AP_Math/control.cpp:417]: don't we risk divide by zero here if input is 0?

**Review by Gone4Dirt — COMMENTED** [review]
Really pleased the change from p gain to rates here.  Will save me getting my calculator our in the feild :-)

**Comment by andyp1per** [comment]
Ok, I flew this. Flies well - no crashes 😄 

- First tried 450/0.4 - this flew pretty much like my old settings of 8/0.4 - good control and enough pop to do rolls and flips
- Next 670/0.6 - this was amazing - really great pop in rolls but still controllable, a little sensitive at mid-stick, maybe 0.7 would be better
- Finally 900/0.8 - this was very flyable at midstick, good control, but the pop was unreal - kept doing double rolls by accident and flips that left me upside down.

So I think this works well and will suit a lot of different pilots

File: ArduCopter/defines.h [hunk 13]
```diff
@@ -53,7 +53,7 @@ enum tuning_func {
     TUNING_ACCEL_Z_KD =                 36, // accel based throttle controller's D term
     TUNING_DECLINATION =                38, // compass declination in radians
     TUNING_CIRCLE_RATE =                39, // circle turn rate in degrees (hard coded to about 45 degrees in either direction)
-    TUNING_ACRO_YAW_KP =                40, // acro controller's P term.  converts pilot input to a desired roll, pitch or yaw rate
+    TUNING_ACRO_YAW_RATE =                40, // acro controller's P term.  converts pilot input to a desired roll, pitch or yaw rate
```
Comment (amilcarlucas) [inline ArduCopter/defines.h:56]: Indent

File: ArduCopter/config.h (hunk 7 above)
Comment (amilcarlucas) [inline ArduCopter/config.h:458]: Would be nice to have the units in a comment, something like:
// [deg/s]

or:

// [rad]

File: ArduCopter/config.h [hunk 14]
```diff
@@ -575,6 +579,13 @@
...
 
+#ifndef PILOT_Y_RATE_DEFAULT
+ # define PILOT_Y_RATE_DEFAULT  202.5   // pilot controlled yaw rotation rate parameter default
+#endif
+#ifndef PILOT_Y_EXPO_DEFAULT
+ # define PILOT_Y_EXPO_DEFAULT  0.0     // pilot controlled yaw expo parameter default
+#endif
+
 #ifndef AUTO_DISARMING_DELAY
 # define AUTO_DISARMING_DELAY  10
```
Comment (amilcarlucas) [inline ArduCopter/config.h:590]: // [s]

File: libraries/AP_Math/control.cpp (hunk 12 above)
Comment (Gone4Dirt) [reply libraries/AP_Math/control.cpp:417]: yeah, my mistake!

Comment (lthall) [reply libraries/AP_Math/control.cpp:417]: I need to add a constrain to ensure the input can't be greater than 1

File: ArduCopter/defines.h (hunk 13 above)
Comment (lthall) [reply ArduCopter/defines.h:56]: And corrected the comments.

File: ArduCopter/config.h (hunk 14 above)
Comment (lthall) [reply ArduCopter/config.h:590]: expo does not have units.

File: libraries/AP_Math/control.cpp (hunk 11 above)
Comment (lthall) [reply libraries/AP_Math/control.cpp:416]: I had to stop it somewhere and I thought a reduction in slope of 95% was reasonable.

File: ArduCopter/Parameters.cpp [hunk 15]
```diff
@@ -429,14 +429,12 @@ const AP_Param::Info Copter::var_info[] = {
     // @Description: Converts pilot roll and pitch into a desired rate of rotation in ACRO and SPORT mode.  Higher values mean faster rate of rotation.
     // @Range: 1 10
     // @User: Standard
```
Comment (amilcarlucas) [inline ArduCopter/Parameters.cpp:431]: Shouldn't you also remove the parameter metadata comments?

File: ArduCopter/Parameters.cpp [hunk 16]
```diff
@@ -429,14 +429,12 @@ const AP_Param::Info Copter::var_info[] = {
...
-    GSCALAR(acro_rp_p,                 "ACRO_RP_P",           ACRO_RP_P),
 
     // @Param: ACRO_YAW_P
     // @DisplayName: Acro Yaw P gain
     // @Description: Converts pilot yaw input into a desired rate of rotation.  Higher values mean faster rate of rotation.
     // @Range: 1 10
```
Comment (amilcarlucas) [inline ArduCopter/Parameters.cpp:436]: Shouldn't you also remove the parameter metadata comments?

File: ArduCopter/config.h [hunk 17]
```diff
@@ -482,6 +478,14 @@
...
+ #define ACRO_RP_RATE_DEFAULT      360 // deg/s
+#endif
+
+#ifndef ACRO_Y_RATE_DEFAULT
+ #define ACRO_Y_RATE_DEFAULT       202.5 // deg/s
+#endif
+
 // RTL Mode
 #ifndef RTL_ALT_FINAL
  # define RTL_ALT_FINAL             0       // the altitude the vehicle will move to as the final stage of Returning to Launch.  Set to zero to land.
```
Comment (amilcarlucas) [inline ArduCopter/config.h:491]: Driveby ... but what are the units here?

File: ArduCopter/mode_drift.cpp [hunk 18]
```diff
@@ -54,7 +54,7 @@ void ModeDrift::run()
 
     // gain scheduling for yaw
     float pitch_vel2 = MIN(fabsf(pitch_vel), 2000);
-    float target_yaw_rate = ((float)target_roll/1.0f) * (1.0f - (pitch_vel2 / 5000.0f)) * g.acro_yaw_p;
+    float target_yaw_rate = target_roll * (1.0f - (pitch_vel2 / 5000.0f)) * g2.acro_y_rate / 45.0;
```
Comment (amilcarlucas) [inline ArduCopter/mode_drift.cpp:57]: Why the `/ 45.0` ?

File: ArduCopter/config.h [hunk 19]
```diff
@@ -575,6 +579,13 @@
  # define PILOT_ACCEL_Z_DEFAULT 250 // vertical acceleration in cm/s/s while altitude is under pilot control
 #endif
 
+#ifndef PILOT_Y_RATE_DEFAULT
+ # define PILOT_Y_RATE_DEFAULT  202.5   // pilot controlled yaw rotation rate parameter default
```
Comment (amilcarlucas) [inline ArduCopter/config.h:583]: deg/s ?

**Review by amilcarlucas — CHANGES_REQUESTED** [review]

File: ArduCopter/mode_drift.cpp (hunk 18 above)
Comment (Gone4Dirt) [reply ArduCopter/mode_drift.cpp:57]: I believe this stems from the output to AP_Motors being an angle which is +/- 4500 cdeg.  So now that we are going to a rate param instead of a gain, we need to convert it back to a gain for equivilent behaviour here.  Looks correct to me.

File: ArduCopter/Parameters.cpp (hunk 15 above)
Comment (Gone4Dirt) [reply ArduCopter/Parameters.cpp:431]: Agreed, we should remove old param metadata.

File: ArduCopter/Parameters.cpp (hunk 16 above)
Comment (Gone4Dirt) [reply ArduCopter/Parameters.cpp:436]: Agreed, we should remove old param metadata.

File: libraries/AP_Math/control.cpp (hunk 11 above)
Comment (Gone4Dirt) [reply libraries/AP_Math/control.cpp:417]: To be fair the 0.95 to 0.99 would probably not be requests as that is a rather extreme expo...

File: ArduCopter/mode.cpp [hunk 20]
```diff
@@ -918,30 +918,19 @@ Mode::AltHoldModeState Mode::get_alt_hold_state(float target_climb_rate_cms)
...
-        float y_in, y_in3, y_out;
 
-        // yaw expo
-        y_in = float(stick_angle)/ROLL_PITCH_YAW_INPUT_MAX;
-        y_in3 = y_in*y_in*y_in;
-        y_out = (g2.acro_y_expo * y_in3) + ((1.0f - g2.acro_y_expo) * y_in);
-        yaw_request = ROLL_PITCH_YAW_INPUT_MAX * y_out * g.acro_yaw_p;
-    }
+    // yaw expo
+    yaw_request = g2.pilot_y_rate * 100.0 * input_expo(yaw_in, g2.pilot_y_expo);
```
Comment (Gone4Dirt) [inline ArduCopter/mode.cpp:932]: we can simplify this to the below and remove the yaw_request float all together.
```suggestion
   return  g2.pilot_y_rate * 100.0 * input_expo(yaw_in, g2.pilot_y_expo);
```

File: libraries/AP_Math/control.cpp [hunk 21]
```diff
@@ -407,3 +407,16 @@ float kinematic_limit(Vector3f direction, float max_xy, float max_z_pos, float m
...
+
+// calculate the expo function on the normalised input
+// input must be in the range of -1 to 1
+// expo should be less than 1.0 but limited to be less than 0.95
+float input_expo(float input, float expo)
+{
+    input = constrain_float(input, -1.0, 1.0);
+    if (expo < 0.95) {
+        return (1 - expo) * input / (1 - expo * fabsf(input));
+    } else {
```
Comment (Gone4Dirt) [inline libraries/AP_Math/control.cpp:419]: nitpick: because we return early if expo < 0.95 we don't need the else {}.  we can just return input; after the if

File: ArduCopter/mode_drift.cpp (hunk 18 above)
Comment (amilcarlucas) [reply ArduCopter/mode_drift.cpp:57]: Please add that to a comment in the code.

File: ArduCopter/Parameters.cpp (hunk 15 above)
Comment (lthall) [reply ArduCopter/Parameters.cpp:431]: I understand that Randy leaves them in. I will see what he says.

File: ArduCopter/Parameters.cpp (hunk 16 above)
Comment (lthall) [reply ArduCopter/Parameters.cpp:436]: I understand that Randy leaves them in. I will see what he says.

File: ArduCopter/config.h (hunk 17 above)
Comment (lthall) [reply ArduCopter/config.h:491]: RTL_ALT_FINAL? cm

File: ArduCopter/mode_drift.cpp (hunk 18 above)
Comment (lthall) [reply ArduCopter/mode_drift.cpp:57]: mode drift is not my code so I can't comment why it is that way. If I had my way i would just delete the mode.

Comment (IamPete1) [reply ArduCopter/mode_drift.cpp:57]: I'm not really a fan either, I suspect we could persuade @rmackay9 ....

File: ArduCopter/Parameters.cpp (hunk 16 above)
Comment (IamPete1) [reply ArduCopter/Parameters.cpp:436]: its safe to delete now we have param versioning on the wiki I think.

**Comment by IamPete1** [comment]
I had not fully realized that this has two sets of yaw rate params. I wonder if its worth it to have both `PILOT` and `ACRO`. I think the move to `PILOT` is a good one. I would be tempted to use them in acro too. Or at least default to 0 to inherit from `PILOT.`

**Comment by lthall** [comment]
> I had not fully realized that this has two sets of yaw rate params. I wonder if its worth it to have both `PILOT` and `ACRO`. I think the move to `PILOT` is a good one. I would be tempted to use them in acro too. Or at least default to 0 to inherit from `PILOT.`

ACRO has unique requirements to all other modes so I a confident that separating them and not defaulting from one to the other is the right way to go. ACRO is a very different pilot interface to Stabilize and I don't want to add restrictions to our future selves by adding relationships to other modes that are not needed.

File: ArduCopter/mode.cpp [hunk 22]
```diff
@@ -918,32 +918,15 @@ Mode::AltHoldModeState Mode::get_alt_hold_state(float target_climb_rate_cms)
 
 // transform pilot's yaw input into a desired yaw rate
 // returns desired yaw rate in centi-degrees per second
-float Mode::get_pilot_desired_yaw_rate(int16_t stick_angle)
+float Mode::get_pilot_desired_yaw_rate(float yaw_in)
```
[the remark this answers is not in the document]
Comment (lthall) [reply ArduCopter/mode.cpp:921]: I think this change is outside the scope of this PR.

File: ArduCopter/defines.h [hunk 23]
```diff
@@ -43,7 +43,7 @@ enum tuning_func {
     TUNING_ALTITUDE_HOLD_KP =           14, // altitude hold controller's P term (alt error to desired rate)
     TUNING_RATE_ROLL_PITCH_KD =         21, // body frame roll/pitch rate controller's D term
     TUNING_VEL_XY_KP =                  22, // loiter rate controller's P term (speed error to tilt angle)
-    TUNING_ACRO_RP_KP =                 25, // acro controller's P term.  converts pilot input to a desired roll, pitch or yaw rate
+    TUNING_ACRO_RP_RATE =               25, // acro controller's desired roll and pitch rate in deg/s
```
Comment (rmackay9) [inline ArduCopter/defines.h:46]: this is slightly dangerous to re-use the old tuning value because the valid range is very different (I think).  I wonder if perhaps we should just remove the ability to tune this parameter at all.  Is it really necessary to be able to change the value with a tuning knob?  Perhaps it was more necessary previously when the values were impossible to understand.

File: ArduCopter/mode.cpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (rmackay9) [reply ArduCopter/mode.cpp:921]: I generally agree with @IamPete1 that the argument is unnecessary.  If we leave it in place though let's put a comment about what the range of the input is.  So I guess it is -1 to +1.

File: ArduCopter/mode.cpp [hunk 24]
```diff
@@ -918,32 +918,15 @@ Mode::AltHoldModeState Mode::get_alt_hold_state(float target_climb_rate_cms)
...
-
-        // yaw expo
-        y_in = float(stick_angle)/ROLL_PITCH_YAW_INPUT_MAX;
-        y_in3 = y_in*y_in*y_in;
-        y_out = (g2.acro_y_expo * y_in3) + ((1.0f - g2.acro_y_expo) * y_in);
-        yaw_request = ROLL_PITCH_YAW_INPUT_MAX * y_out * g.acro_yaw_p;
-    }
     // convert pilot input to the desired yaw rate
-    return yaw_request;
+    return g2.pilot_y_rate * 100.0 * input_expo(yaw_in, g2.pilot_y_expo);
```
Comment (rmackay9) [inline ArduCopter/mode.cpp:929]: No need for protection against crazy values?  I'm happy either way but I just want to be sure we've thought about this.  For example is zero ok?

File: ArduCopter/mode.h [hunk 25]
```diff
@@ -315,7 +315,7 @@ class ModeAcro : public Mode {
     const char *name() const override { return "ACRO"; }
     const char *name4() const override { return "ACRO"; }
 
-    void get_pilot_desired_angle_rates(int16_t roll_in, int16_t pitch_in, int16_t yaw_in, float &roll_out, float &pitch_out, float &yaw_out);
+    void get_pilot_desired_angle_rates(float roll_in, float pitch_in, float yaw_in, float &roll_out, float &pitch_out, float &yaw_out);
```
Comment (rmackay9) [inline ArduCopter/mode.h:318]: Because we've touched the method let's put a comment above it to correct for past wrongs.  That comment should mention the range of the arguments because the argument names themselves don't make it at all clear.

File: ArduCopter/mode_acro.cpp [hunk 26]
```diff
@@ -96,46 +96,30 @@ float ModeAcro::throttle_hover() const
 
 // get_pilot_desired_angle_rates - transform pilot's roll pitch and yaw input into a desired lean angle rates
 // returns desired angle rates in centi-degrees-per-second
-void ModeAcro::get_pilot_desired_angle_rates(int16_t roll_in, int16_t pitch_in, int16_t yaw_in, float &roll_out, float &pitch_out, float &yaw_out)
+void ModeAcro::get_pilot_desired_angle_rates(float roll_in, float pitch_in, float yaw_in, float &roll_out, float &pitch_out, float &yaw_out)
```
Comment (rmackay9) [inline ArduCopter/mode_acro.cpp:99]: Let's put a comment here too specifying what the range of the roll_in, pitch_in and yaw_in arguments.. or make it clear by renaming the arguments.

File: ArduCopter/mode_drift.cpp (hunk 18 above)
Comment (rmackay9) [reply ArduCopter/mode_drift.cpp:57]: I'm certainly a fan of simplifying Copter to remove features that almost nobody uses.  Undoubtedly someone will turn up and say they use the mode (it came up once during 4.1 beta testing) but we could argue for the sake of supportability and progress that it should go.  The first step would be to make it a compiled-off-by-default option and then a version later we remove it assuming there haven't been too many complaints.

File: ArduCopter/mode_acro.cpp [hunk 27]
```diff
@@ -96,46 +96,30 @@ float ModeAcro::throttle_hover() const
...
-
-        // pitch expo
-        rp_in = float(pitch_in)/ROLL_PITCH_YAW_INPUT_MAX;
-        rp_in3 = rp_in*rp_in*rp_in;
-        rp_out = (g.acro_rp_expo * rp_in3) + ((1.0f - g.acro_rp_expo) * rp_in);
-        rate_bf_request.y = ROLL_PITCH_YAW_INPUT_MAX * rp_out * g.acro_rp_p;
-    }
+    
+    // roll expo
+    rate_bf_request_cd.x = g2.acro_rp_rate * 100.0 * input_expo(roll_in, g.acro_rp_expo);
```
Comment (rmackay9) [inline ArduCopter/mode_acro.cpp:116]: no need for any protection against invalid values?

File: ArduCopter/mode.cpp (hunk 24 above)
Comment (rmackay9) [reply ArduCopter/mode.cpp:929]: actually I think this is OK actually..

File: ArduCopter/defines.h (hunk 23 above)
Comment (rmackay9) [reply ArduCopter/defines.h:46]: OK, Leonard wants to leave this ability in so it's ok.  It's a bit of an edge case to worry about this.

So we should update the TUNE parameter description.

**Comment by rmackay9** [comment]
@IamPete1, re putting some of these parameters into a library, this is not a new idea actually.  Rob Lefebvre had this idea many years ago and create https://github.com/ArduPilot/ardupilot/tree/master/libraries/AC_InputManager.

File: ArduCopter/defines.h (hunk 23 above)
Comment (lthall) [reply ArduCopter/defines.h:46]: Done

File: ArduCopter/mode_acro.cpp (hunk 27 above)
Comment (lthall) [reply ArduCopter/mode_acro.cpp:116]: Protection is contained in the input_expo function.

File: ArduCopter/mode.cpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (lthall) [reply ArduCopter/mode.cpp:921]: I am told this is a line ball with a leaning towards leaving it in for this PR.

File: ArduCopter/Parameters.cpp (hunk 15 above)
Comment (rmackay9) [reply ArduCopter/Parameters.cpp:431]: Last time I checked MP beta used param descriptions from "latest" while MP stable used param descriptions from ArduPilot stable (i.e. flight code stable). I'm unsure if this has changed since but let's leave the older param descriptions in place until we know.

File: ArduCopter/Parameters.cpp (hunk 16 above)
Comment (rmackay9) [reply ArduCopter/Parameters.cpp:436]: let's leave it in until we have confirmation from MP.

**Review by rmackay9 — APPROVED** [review]
